lib_eio_luv/eio_luv.ml

```diff
@@ -575,10 +575,15 @@ module Low_level = struct
       | _ -> None
     in
     let request = Luv.DNS.Addr_info.Request.make () in
-    await_with_cancel ~request (fun loop -> Luv.DNS.getaddrinfo ~loop ~request ~service ~node ())
-    |> or_raise
-    |> List.filter_map to_eio_sockaddr_t
-
+    match (await_with_cancel ~request
+             (fun loop -> Luv.DNS.getaddrinfo ~loop ~request ~service ~node ()))
+    with
+    | Ok nl -> List.filter_map to_eio_sockaddr_t nl
+    | Error `EAI_ADDRFAMILY | Error `EAI_AGAIN | Error `EAI_BADFLAGS | Error `EAI_BADHINTS
+    | Error `EAI_CANCELED   | Error `EAI_FAIL  | Error `EAI_FAMILY   | Error `EAI_MEMORY
+    | Error `EAI_NODATA     | Error `EAI_NONAME| Error `EAI_OVERFLOW | Error `EAI_PROTOCOL
+    | Error `EAI_SERVICE    | Error `EAI_SOCKTYPE -> []
```

Comment:
This list seems excessive (even if it does match the stdlib version). In particular, `EAI_AGAIN`, `EAI_BADFLAGS`, `EAI_MEMORY` and `EAI_SOCKTYPE` seem like they should be reported as errors, rather than claiming there are no addresses. Maybe `EAI_FAIL` too.

Reply:
I thought about it too, just some notes:
EAI_FAIL and EAI_AGAIN can be the result of a an actual DNS reply (would have to dig deeper, but those should return an empty list). EAI_SOCKTYPE, EAI_BADFLAGS, EAI_FAMILY we could, since they are mostly programming errors from our side.
I don't have a strong opinion, but I'd err on the side of just emulating the same behaviour of `Unix.getaddrinfo` which ignores everything.

Reply:
I wonder if we should just copy the `getaddrinfo` API and not allow an empty result. The wrapper function could enforce this:

```ocaml
let getaddrinfo ?(service="") (t:#t) hostname =
  match t#getaddrinfo ~service hostname with
  | [] -> raise (Dns_error (Failure "No addresses for ..."))
  | xs -> xs
```

I guess most people getting an empty list will want to report an error, and it's a bit silly you have to guess what the error was when we just threw the real error away. e.g. in `with_tcp_connect` we currently do:

```ocaml
    | [] -> raise (Connection_failure (Failure (Fmt.str "No TCP addresses for %S" host)))
```
Then we could say if that's because the DNS server is busy, the entry doesn't exist, doesn't have an IP address, etc.

Reply:
Also `Dns_error` is a bit imprecise since `getaddrinfo` is not only about DNS (`AI_NUMERICHOST`and so on)

Reply:
Thinking a bit more about this, I think `connect` should take a `Sockaddr.stream list`, and loop over the candidates until it succeeds connecting to one. This is virtually what everyone does with `getaddrinfo` + `connect` and `bind`.
In this case `connect` itself could raise something, `getaddrinfo` could still return just an empty list, and the user would still have to handle an exception.

Reply:
Yes, the error code was just an example. We can use a better exception name and contents.

> I think connect should take a Sockaddr.stream list, and loop over the candidates until it succeeds connecting to one.

`with_tcp_connect` already handles looping over the candidates; I'm not sure there's much benefit to changing `connect` too. In fact, that would just mean having to do something for the extra case of an empty list.

> In this case connect itself could raise something, getaddrinfo could still return just an empty list, and the user would still have to handle an exception.

I don't see how that helps. We'd still be losing the detailed error from `getaddrinfo` and replacing it with a guess.

Reply:
can I get a ruling here ?

Reply:
> Should we just raise all errors from `getaddrinfo` on both cases?

Yes, that makes sense to me.

( Having `connect` take a `Sockaddr.stream list` doesn't seem useful to me. You can't give a good error if there are no addresses - you don't know what the DNS error was, what the failing hostname was, or even whether DNS was used at all! )